# HG changeset patch
# Parent 936a69ffd85363a9cb428f9011bafafb6f74bd3e
# Parent  f9aa8a3fd77288a69e0be23f0623ce31868bab19
Add the ability to target a specific device via its address

diff --git a/compat/generic/time_monotonic.h b/compat/generic/time_monotonic.h
deleted file mode 100644
--- a/compat/generic/time_monotonic.h
+++ /dev/null
@@ -1,34 +0,0 @@
-// Copyright (c) 2015, Louis Opter <kalessin@kalessin.fr>
-// All rights reserved.
-//
-// Redistribution and use in source and binary forms, with or without
-// modification, are permitted provided that the following conditions are met:
-//
-// 1. Redistributions of source code must retain the above copyright notice,
-//    this list of conditions and the following disclaimer.
-//
-// 2. Redistributions in binary form must reproduce the above copyright notice,
-//    this list of conditions and the following disclaimer in the documentation
-//    and/or other materials provided with the distribution.
-//
-// 3. Neither the name of the copyright holder nor the names of its contributors
-//    may be used to endorse or promote products derived from this software
-//    without specific prior written permission.
-//
-// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
-// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
-// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
-// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
-// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
-// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
-// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
-// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
-// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
-// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
-// POSSIBILITY OF SUCH DAMAGE.
-
-#pragma once
-
-typedef time_t lgtd_time_mono_t;
-
-lgtd_time_mono_t lgtd_time_monotonic_msecs(void);
diff --git a/core/router.c b/core/router.c
--- a/core/router.c
+++ b/core/router.c
@@ -79,6 +79,33 @@
     }
 }
 
+static void
+lgtd_router_device(struct lgtd_lifx_bulb *bulb,
+                   enum lgtd_lifx_packet_type pkt_type,
+                   void *pkt)
+{
+    assert(bulb);
+
+    struct lgtd_lifx_packet_header hdr;
+    union lgtd_lifx_target target = { .addr = bulb->addr };
+
+    const struct lgtd_lifx_packet_infos *pkt_infos;
+    pkt_infos = lgtd_lifx_wire_setup_header(
+        &hdr, LGTD_LIFX_TARGET_DEVICE, target, bulb->gw->site, pkt_type
+    );
+    assert(pkt_infos);
+
+    lgtd_lifx_gateway_send_packet(bulb->gw, &hdr, pkt, pkt_infos->size);
+
+    if (pkt_type == LGTD_LIFX_SET_POWER_STATE) {
+        bulb->dirty_at = lgtd_time_monotonic_msecs();
+        struct lgtd_lifx_packet_power_state *payload = pkt;
+        bulb->expected_power_on = payload->power;
+    }
+
+    lgtd_debug("sending %s to %s", pkt_infos->name, lgtd_addrtoa(bulb->addr));
+}
+
 bool
 lgtd_router_send(const char *target,
                  enum lgtd_lifx_packet_type pkt_type,
@@ -91,5 +118,25 @@
         return true;
     }
 
+    if (isxdigit(target[0])) {
+        const char *endptr = NULL;
+        errno = 0;
+        long long device = strtoll(target, (char **)&endptr, 16);
+        if (*endptr || errno == ERANGE) {
+            lgtd_debug("invalid target device %s", target);
+            return false;
+        }
+        device = htobe64(device);
+        struct lgtd_lifx_bulb *bulb = lgtd_lifx_bulb_get(
+            (uint8_t *)&device + sizeof(device) - LGTD_LIFX_ADDR_LENGTH
+        );
+        if (!bulb) {
+            lgtd_debug("target device %#llx not found", device);
+            return false;
+        }
+        lgtd_router_device(bulb, pkt_type, pkt);
+        return true;
+    }
+
     return false;
 }
diff --git a/lifx/bulb.c b/lifx/bulb.c
--- a/lifx/bulb.c
+++ b/lifx/bulb.c
@@ -49,9 +49,8 @@
     RB_INITIALIZER(&lgtd_lifx_bulbs_table);
 
 struct lgtd_lifx_bulb *
-lgtd_lifx_bulb_get(struct lgtd_lifx_gateway *gw, const uint8_t *addr)
+lgtd_lifx_bulb_get(const uint8_t *addr)
 {
-    assert(gw);
     assert(addr);
 
     struct lgtd_lifx_bulb bulb;
@@ -89,9 +88,10 @@
     RB_REMOVE(lgtd_lifx_bulb_map, &lgtd_lifx_bulbs_table, bulb);
     SLIST_REMOVE(&bulb->gw->bulbs, bulb, lgtd_lifx_bulb, link_by_gw);
     lgtd_info(
-        "closed bulb \"%.*s\" on [%s]:%hu",
+        "closed bulb \"%.*s\" (%s) on [%s]:%hu",
         LGTD_LIFX_LABEL_SIZE,
         bulb->state.label,
+        lgtd_addrtoa(bulb->addr),
         bulb->gw->ip_addr,
         bulb->gw->port
     );
diff --git a/lifx/bulb.h b/lifx/bulb.h
--- a/lifx/bulb.h
+++ b/lifx/bulb.h
@@ -72,7 +72,7 @@
     lgtd_lifx_bulb_cmp
 );
 
-struct lgtd_lifx_bulb *lgtd_lifx_bulb_get(struct lgtd_lifx_gateway *, const uint8_t *);
+struct lgtd_lifx_bulb *lgtd_lifx_bulb_get(const uint8_t *);
 struct lgtd_lifx_bulb *lgtd_lifx_bulb_open(struct lgtd_lifx_gateway *, const uint8_t *);
 void lgtd_lifx_bulb_close(struct lgtd_lifx_bulb *);
 
diff --git a/lifx/gateway.c b/lifx/gateway.c
--- a/lifx/gateway.c
+++ b/lifx/gateway.c
@@ -166,7 +166,7 @@
     assert(gw);
     assert(bulb_addr);
 
-    struct lgtd_lifx_bulb *bulb = lgtd_lifx_bulb_get(gw, bulb_addr);
+    struct lgtd_lifx_bulb *bulb = lgtd_lifx_bulb_get(bulb_addr);
     if (!bulb) {
         bulb = lgtd_lifx_bulb_open(gw, bulb_addr);
         if (bulb) {
diff --git a/tests/core/jsonrpc/CMakeLists.txt b/tests/core/jsonrpc/CMakeLists.txt
--- a/tests/core/jsonrpc/CMakeLists.txt
+++ b/tests/core/jsonrpc/CMakeLists.txt
@@ -1,7 +1,9 @@
 INCLUDE_DIRECTORIES(
     ${LIGHTSD_SOURCE_DIR}
+    ${LIGHTSD_SOURCE_DIR}/../
     ${LIGHTSD_SOURCE_DIR}/core/
     ${LIGHTSD_BINARY_DIR}
+    ${LIGHTSD_BINARY_DIR}/../
     ${LIGHTSD_BINARY_DIR}/core/
 )
 
@@ -9,7 +11,8 @@
     test_core_jsonrpc STATIC
     ${LIGHTSD_SOURCE_DIR}/core/jsmn.c
     ${LIGHTSD_SOURCE_DIR}/core/log.c
-    ${CMAKE_CURRENT_SOURCE_DIR}/../utils.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/../tests_shims.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/../tests_utils.c
 )
 
 FUNCTION(ADD_JSONRPC_TEST TEST_SOURCE)
diff --git a/tests/core/router/CMakeLists.txt b/tests/core/router/CMakeLists.txt
new file mode 100644
--- /dev/null
+++ b/tests/core/router/CMakeLists.txt
@@ -0,0 +1,33 @@
+INCLUDE_DIRECTORIES(
+    ${LIGHTSD_SOURCE_DIR}
+    ${LIGHTSD_SOURCE_DIR}/../
+    ${LIGHTSD_SOURCE_DIR}/core/
+    ${LIGHTSD_SOURCE_DIR}/tests/core/
+    ${LIGHTSD_BINARY_DIR}
+    ${LIGHTSD_BINARY_DIR}/../
+    ${LIGHTSD_BINARY_DIR}/core/
+    ${LIGHTSD_BINARY_DIR}/tests/core/
+)
+
+ADD_LIBRARY(
+    test_core_router STATIC
+    ${LIGHTSD_SOURCE_DIR}/core/log.c
+    ${LIGHTSD_SOURCE_DIR}/core/proto.c
+    ${LIGHTSD_SOURCE_DIR}/lifx/bulb.c
+    ${LIGHTSD_SOURCE_DIR}/lifx/timer.c
+    ${LIGHTSD_SOURCE_DIR}/lifx/wire_proto.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/../tests_shims.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/../tests_utils.c
+    ${TIME_MONOTONIC_IMPL}
+)
+
+TARGET_LINK_LIBRARIES(test_core_router ${EVENT2_CORE_LIBRARY})
+
+FUNCTION(ADD_ROUTER_TEST TEST_SOURCE)
+    ADD_TEST_FROM_C_SOURCES(${TEST_SOURCE} test_core_router)
+ENDFUNCTION()
+
+FILE(GLOB TESTS RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "test_*.c")
+FOREACH(TEST ${TESTS})
+    ADD_ROUTER_TEST(${TEST})
+ENDFOREACH()
diff --git a/tests/core/router/test_router_broadcast.c b/tests/core/router/test_router_broadcast.c
new file mode 100644
--- /dev/null
+++ b/tests/core/router/test_router_broadcast.c
@@ -0,0 +1,48 @@
+#include "router.c"
+
+#include "tests_utils.h"
+#include "tests_router_utils.h"
+
+int
+main(void)
+{
+    lgtd_lifx_wire_load_packet_infos_map();
+
+    lgtd_tests_insert_mock_gateway(2);
+    lgtd_tests_insert_mock_gateway(1);
+
+    struct lgtd_lifx_packet_power_state payload = {
+        .power = LGTD_LIFX_POWER_ON
+    };
+    lgtd_router_send("*", LGTD_LIFX_SET_POWER_STATE, &payload);
+
+    if (lgtd_gw_pkt_queue_size != 2) {
+        lgtd_errx(1, "2 packets should have been sent");
+    }
+
+    for (int i = lgtd_gw_pkt_queue_size; i--;) {
+        struct lgtd_lifx_gateway *gw = lgtd_gw_pkt_queue[i].gw;
+        if (gw->socket != i + 1) {
+            lgtd_errx(
+                1, "packet was sent to wrong gateway (expected %d, got %d)",
+                i + 1, gw->socket
+            );
+        }
+        const struct lgtd_lifx_packet_header *hdr;
+        hdr = lgtd_gw_pkt_queue[i].hdr;
+        if (!hdr->protocol.tagged || hdr->protocol.addressable) {
+            lgtd_errx(1, "packet header doesn't have the right bits set");
+        }
+        if (hdr->target.tags != 0) {
+            lgtd_errx(1, "tags should be 0 for broadcast");
+        }
+        if (memcmp(gw->site, hdr->site, sizeof(hdr->site))) {
+            lgtd_errx(1, "sites don't match");
+        }
+        if (lgtd_gw_pkt_queue[i].pkt != &payload) {
+            lgtd_errx(1, "the payload has been improperly set");
+        }
+    }
+
+    return 0;
+}
diff --git a/tests/core/router/tests_router_utils.h b/tests/core/router/tests_router_utils.h
new file mode 100644
--- /dev/null
+++ b/tests/core/router/tests_router_utils.h
@@ -0,0 +1,39 @@
+#include <sys/queue.h>
+#include <sys/tree.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <event2/util.h>
+
+#include "lifx/wire_proto.h"
+#include "core/time_monotonic.h"
+#include "lifx/bulb.h"
+#include "lifx/gateway.h"
+#include "tests_utils.h"
+
+int lgtd_gw_pkt_queue_size = 0;
+struct {
+    struct lgtd_lifx_gateway        *gw;
+    struct lgtd_lifx_packet_header  *hdr;
+    const void                      *pkt;
+    int                             pkt_size;
+} lgtd_gw_pkt_queue[16] = { { NULL, NULL, NULL, 0}, };
+
+void
+lgtd_lifx_gateway_send_packet(struct lgtd_lifx_gateway *gw,
+                              const struct lgtd_lifx_packet_header *hdr,
+                              const void *pkt,
+                              int pkt_size)
+{
+    lgtd_gw_pkt_queue[lgtd_gw_pkt_queue_size].gw = gw;
+    // headers are created on the stack so we need to dup them:
+    lgtd_gw_pkt_queue[lgtd_gw_pkt_queue_size].hdr = malloc(sizeof(*hdr));
+    memcpy(lgtd_gw_pkt_queue[lgtd_gw_pkt_queue_size].hdr, hdr, sizeof(*hdr));
+    lgtd_gw_pkt_queue[lgtd_gw_pkt_queue_size].pkt = pkt;
+    lgtd_gw_pkt_queue[lgtd_gw_pkt_queue_size].pkt_size = pkt_size;
+    lgtd_gw_pkt_queue_size++;
+}
diff --git a/tests/core/tests_shims.c b/tests/core/tests_shims.c
new file mode 100644
--- /dev/null
+++ b/tests/core/tests_shims.c
@@ -0,0 +1,58 @@
+#include <sys/queue.h>
+#include <sys/tree.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <event2/event.h>
+#include <event2/util.h>
+
+#include "lifx/wire_proto.h"
+#include "core/time_monotonic.h"
+#include "lifx/bulb.h"
+#include "lifx/gateway.h"
+#include "lightsd.h"
+
+struct lgtd_opts lgtd_opts = {
+    .foreground = false,
+    .log_timestamps = false,
+    .verbosity = LGTD_DEBUG
+};
+
+struct event_base *lgtd_ev_base = NULL;
+
+void
+lgtd_cleanup(void)
+{
+}
+
+
+void lgtd_lifx_gateway_handle_pan_gateway(struct lgtd_lifx_gateway *gw,
+                                          const struct lgtd_lifx_packet_header *hdr,
+                                          const struct lgtd_lifx_packet_pan_gateway *pkt)
+{
+    (void)gw;
+    (void)hdr;
+    (void)pkt;
+}
+
+void lgtd_lifx_gateway_handle_light_status(struct lgtd_lifx_gateway *gw,
+                                           const struct lgtd_lifx_packet_header *hdr,
+                                           const struct lgtd_lifx_packet_light_status *pkt)
+{
+    (void)gw;
+    (void)hdr;
+    (void)pkt;
+}
+
+void lgtd_lifx_gateway_handle_power_state(struct lgtd_lifx_gateway *gw,
+                                          const struct lgtd_lifx_packet_header *hdr,
+                                          const struct lgtd_lifx_packet_power_state *pkt)
+{
+    (void)gw;
+    (void)hdr;
+    (void)pkt;
+}
diff --git a/tests/core/utils.c b/tests/core/tests_utils.c
rename from tests/core/utils.c
rename to tests/core/tests_utils.c
--- a/tests/core/utils.c
+++ b/tests/core/tests_utils.c
@@ -1,18 +1,32 @@
+#include <sys/queue.h>
+#include <sys/tree.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
 #include <stdbool.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
 
-#include <event2/event.h>
+#include <event2/util.h>
 
-#include "lightsd.h"
+#include "lifx/wire_proto.h"
+#include "core/time_monotonic.h"
+#include "lifx/bulb.h"
+#include "lifx/gateway.h"
+#include "tests_utils.h"
 
-struct lgtd_opts lgtd_opts = {
-    .foreground = false,
-    .log_timestamps = false,
-    .verbosity = LGTD_DEBUG
-}; 
+struct lgtd_lifx_gateway_list lgtd_lifx_gateways =
+    LIST_HEAD_INITIALIZER(&lgtd_lifx_gateways);
 
-struct event_base *lgtd_ev_base = NULL;
+struct lgtd_lifx_gateway *
+lgtd_tests_insert_mock_gateway(int id)
+{
+    struct lgtd_lifx_gateway *gw = calloc(1, sizeof(*gw));
 
-void
-lgtd_cleanup(void)
-{
+    gw->socket = id;
+    gw->site[0] = id;
+
+    LIST_INSERT_HEAD(&lgtd_lifx_gateways, gw, link);
+
+    return gw;
 }
diff --git a/tests/core/tests_utils.h b/tests/core/tests_utils.h
new file mode 100644
--- /dev/null
+++ b/tests/core/tests_utils.h
@@ -0,0 +1,1 @@
+struct lgtd_lifx_gateway *lgtd_tests_insert_mock_gateway(int);
diff --git a/tests/lightsc b/tests/lightsc
--- a/tests/lightsc
+++ b/tests/lightsc
@@ -18,30 +18,31 @@
     print(response)
 
 
-def set_light_from_hsbk(socket, id, h, s, b, k):
+def set_light_from_hsbk(socket, id, target, h, s, b, k):
     jsonrpc_call(socket, id, "set_light_from_hsbk", [
-        "*", h, s, b, k
+        target, h, s, b, k
     ])
 
 
-def power_on(socket, id):
-    jsonrpc_call(socket, id, "power_on", {"target": "*"})
+def power_on(socket, id, target):
+    jsonrpc_call(socket, id, "power_on", {"target": target})
 
 
-def power_off(socket, id):
-    jsonrpc_call(socket, id, "power_off", {"target": "*"})
+def power_off(socket, id, target):
+    jsonrpc_call(socket, id, "power_off", {"target": target})
 
 if __name__ == "__main__":
     s = socket.create_connection(("localhost", 1234))
     h = 0
     id = 0
+    target = "*"  # "d073d500603b"
     try:
-        power_on(s, id)
+        power_on(s, id, target)
         while True:
             h = (h + 1) % 360
             id += 1
-            set_light_from_hsbk(s, id, h, 0.7, 0.02, 2500)
+            set_light_from_hsbk(s, id, target, h, 0.7, 0.02, 2500)
             time.sleep(0.1)
     finally:
-        power_off(s, id)
+        power_off(s, id, target)
         s.close()
